File: serp_llm/src/infrastructure/db/models/message_model.py
# Author: QuanTuanHuy
# Description: Part of Serp Project - Message Database Model

from typing import TYPE_CHECKING, Optional
from sqlalchemy import BigInteger, String, Text, Integer, JSON, ForeignKey, Index
from sqlalchemy.orm import Mapped, mapped_column, relationship

from src.infrastructure.db.base import Base
from src.infrastructure.db.models.base_model import BaseModel

if TYPE_CHECKING:
    from src.infrastructure.db.models.conversation_model import ConversationModel


class MessageModel(Base, BaseModel):
    """Message model for database persistence"""
    
    __tablename__ = "messages"
    
    conversation_id: Mapped[int] = mapped_column(
        BigInteger,
        ForeignKey("conversations.id", ondelete="CASCADE"),
        nullable=False,
        index=True
    )
    
    role: Mapped[str] = mapped_column(String(20), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    tokens_used: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    
    meta_data: Mapped[dict] = mapped_column(JSON, default=dict, nullable=False)
    
    conversation: Mapped["ConversationModel"] = relationship(
        "ConversationModel",
        back_populates="messages"
    )
    
    __table_args__ = (
        Index("idx_conversation_id", "conversation_id"),
    )
    
    def __repr__(self) -> str:
        return f"<MessageModel(id={self.id}, conversation_id={self.conversation_id}, role={self.role})>"

File: discuss_service/src/main/java/serp/project/discuss_service/infrastructure/store/model/MessageModel.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project - Message JPA model
 */

package serp.project.discuss_service.infrastructure.store.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;
import serp.project.discuss_service.core.domain.enums.MessageType;

import java.util.List;
import java.util.Map;

@Entity
@Table(name = "messages", indexes = {
    @Index(name = "idx_messages_channel_created", columnList = "channel_id, created_at DESC"),
    @Index(name = "idx_messages_parent", columnList = "parent_id"),
    @Index(name = "idx_messages_sender", columnList = "sender_id")
})
@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@SuperBuilder
public class MessageModel extends BaseModel {

    @Column(name = "channel_id", nullable = false)
    private Long channelId;

    @Column(name = "sender_id", nullable = false)
    private Long senderId;

    @Column(name = "tenant_id", nullable = false)
    private Long tenantId;

    @Column(name = "content", columnDefinition = "TEXT")
    private String content;

    @Column(name = "message_type", nullable = false, length = 20)
    @Enumerated(EnumType.STRING)
    private MessageType messageType;

    @JdbcTypeCode(SqlTypes.ARRAY)
    @Column(name = "mentions", columnDefinition = "BIGINT[]")
    private Long[] mentions;

    @Column(name = "parent_id")
    private Long parentId;

    @Column(name = "thread_count", nullable = false)
    private Integer threadCount = 0;

    @Column(name = "is_edited", nullable = false)
    private Boolean isEdited = false;

    @Column(name = "edited_at")
    private Long editedAt;

    @Column(name = "is_deleted", nullable = false)
    private Boolean isDeleted = false;

    @Column(name = "deleted_at")
    private Long deletedAt;

    @Column(name = "deleted_by")
    private Long deletedBy;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "reactions", columnDefinition = "jsonb")
    private List<Map<String, Object>> reactions;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "metadata", columnDefinition = "jsonb")
    private Map<String, Object> metadata;

    // Full-text search vector (generated column - read only)
    @Column(name = "search_vector", insertable = false, updatable = false)
    private String searchVector;
}
